### PR TITLE
Nullable annotations for CommandSearcher

### PR DESCRIPTION
**Pull request by powercode**
# PR Summary

Nullable annotations for CommandSearcher

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 1]
```diff
@@ -355,13 +358,13 @@ private CommandInfo ProcessQualifiedFileSystemState()
...
 
-            CommandInfo currentMatch = null;
+            CommandInfo? currentMatch = null;
             _currentState = SearchState.PathSearch;
             if (_canDoPathLookup)
             {
                 try
                 {
-                    while (currentMatch == null && _pathSearcher.MoveNext())
+                    while (currentMatch == null && _pathSearcher!.MoveNext())
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/CommandSearcher.cs:367]: can you add a comment explaining why `_pathSearcher` is not null here

File: src/System.Management.Automation/engine/MshSnapinQualifiedName.cs [hunk 2]
```diff
@@ -62,11 +64,11 @@ private PSSnapinQualifiedName(string[] splitName)
         /// <returns>
         /// An instance of the Name class.
         /// </returns>
-        internal static PSSnapinQualifiedName GetInstance(string name)
+        internal static PSSnapinQualifiedName? GetInstance(string name)
         {
             if (name == null)
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/MshSnapinQualifiedName.cs:69]: we can remove this null check

File: src/System.Management.Automation/engine/MshSnapinQualifiedName.cs [hunk 3]
```diff
@@ -62,11 +64,11 @@ private PSSnapinQualifiedName(string[] splitName)
...
                 return null;
-            PSSnapinQualifiedName result = null;
+            PSSnapinQualifiedName? result = null;
```
Comment (xtqqczze) [inline src/System.Management.Automation/engine/MshSnapinQualifiedName.cs:71]: unnecessary initialization, `result` can be made non-nullable:
```suggestion
            PSSnapinQualifiedName result;
```

File: src/System.Management.Automation/engine/MshSnapinQualifiedName.cs (hunk 2 above)
Comment (xtqqczze) [inline src/System.Management.Automation/engine/MshSnapinQualifiedName.cs:69]: we can remove the null check `(name == null)`

Comment (powercode) [reply src/System.Management.Automation/engine/MshSnapinQualifiedName.cs:69]: No, but I'll make name nullable

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 4]
```diff
@@ -870,70 +874,61 @@ private bool ShouldSkipCommandResolutionForConstrainedLanguage(CommandInfo resul
...
                         module = m;
                         break;
                     }
                 }
 
-                if (module == null)
-                {
-                    module = modules[0];
-                }
+                module ??= modules[0];
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandSearcher.cs:931]: Please revert unrelated change.

File: src/System.Management.Automation/engine/CommandPathSearch.cs [hunk 5]
```diff
@@ -1,14 +1,14 @@
 // Copyright (c) Microsoft Corporation.
```
Comment (PoshChan) [inline src/System.Management.Automation/engine/CommandPathSearch.cs:1]: @powercode, your last commit had 1 failures in `PowerShell-CI-macos`
Job Cmdlet Tests.Retrieving partial output from jobs.Receive-Job will retrieve partial output, including -Keep results
```powershell
Destination array was not long enough. Check the destination index, length, and the array's lower bounds. (Parameter 'destinationArray')
at <ScriptBlock>, /Users/runner/runners/2.168.2/work/1/s/test/powershell/Modules/Microsoft.PowerShell.Core/Job.Tests.ps1: line 108
108:                         $results = Receive-Job -Keep $job

```

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 6]
```diff
@@ -1297,7 +1293,7 @@ private string GetNextLiteralPathThatExistsAndHandleExceptions(string command, o
         /// <returns>
         /// Full path to the command.
         /// </returns>
-        private string GetNextLiteralPathThatExists(string command, out ProviderInfo provider)
+        private string? GetNextLiteralPathThatExists(string? command, out ProviderInfo? provider)
         {
             string resolvedPath = _context.LocationGlobber.GetProviderPath(command, out provider);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandSearcher.cs:1298]: It seems we could re-review after annotating LocationGlobber.GetProviderPath()

File: src/System.Management.Automation/engine/CommandSearcher.cs [hunk 7]
```diff
@@ -1495,12 +1491,12 @@ private static CanDoPathLookupResult CanDoPathLookup(string possiblePath)
         /// The enumerator that uses the Path to
         /// search for commands.
         /// </summary>
-        private CommandPathSearch _pathSearcher;
+        private CommandPathSearch? _pathSearcher;
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/CommandSearcher.cs:1494]: Perhaps there is an example in .Net Runtime repo how make this non-nullable.

Comment (powercode) [reply src/System.Management.Automation/engine/CommandSearcher.cs:1494]: That is a feature under development, but not available in the current compiler.

**Review by iSazonov — APPROVED** [review]
